community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/EagerAggregationPipe.scala

```diff
@@ -51,34 +53,59 @@ case class EagerAggregationPipe(source: Pipe, keyExpressions: Set[String], aggre
     state.decorator.registerParentPipe(this)
 
     // This is the temporary storage used while the aggregation is going on
-    val result = MutableMap[NiceHasher, Seq[AggregationFunction]]()
-    val keyNames: Seq[String] = keyExpressions.toSeq
+    val result = MutableMap[Equals, Seq[AggregationFunction]]()
+    val keyNames = keyExpressions.toList
     val aggregationNames: Seq[String] = aggregations.keys.toSeq
-    val mapSize = keyNames.size + aggregationNames.size
+    val keyNamesSize = keyNames.size
+    val mapSize = keyNamesSize + aggregationNames.size
 
-    def createResults(key: NiceHasher, aggregator: scala.Seq[AggregationFunction]): ExecutionContext = {
+    def createEmptyResult(params: Map[String, Any]): Iterator[ExecutionContext] = {
+      val newMap = MutableMaps.empty
+      val aggregationNamesAndFunctions = aggregationNames zip aggregations.map(_._2.createAggregationFunction.result)
+
+      aggregationNamesAndFunctions.toMap
+        .foreach { case (name, zeroValue) => newMap += name -> zeroValue}
+      Iterator.single(ExecutionContext(newMap))
+    }
+
+    // This code is not pretty. It's full of asInstanceOf calls and other things that might irk you.
+    // You'll just have to trust that the original authors spent time profiling and making sure that this
+    // code runs really fast.
+    // If you feel like cleaning it up - please make sure to not regress in performance. This is a hot spot.
+    def createResults(key: Any, aggregator: scala.Seq[AggregationFunction]): ExecutionContext = {
       val newMap = MutableMaps.create(mapSize)
 
       //add key values
-      (keyNames zip key.original).foreach(newMap += _)
+      keyNamesSize match {
+        case 1 =>
+          newMap += keyNames.head -> key.asInstanceOf[Equivalent].originalValue
+        case 2 =>
+          val t2 = key.asInstanceOf[(Equivalent, Equivalent)]
+          newMap += keyNames.head -> t2._1.originalValue +=
+                    keyNames.last -> t2._2.originalValue
+        case 3 =>
+          val t3 = key.asInstanceOf[(Equivalent, Equivalent, Equivalent)]
+          newMap += keyNames.head -> t3._1.originalValue +=
+                    keyNames.tail.head -> t3._2.originalValue +=
+                    keyNames.last -> t3._3.originalValue
+        case _ =>
+          val listOfValues = key.asInstanceOf[List[Equivalent]]
+          (keyNames zip listOfValues.map(_.originalValue)).foreach(newMap += _)
```

Comment:
I understand that this is the reason for doing the same thing three times in the other place. Seems to make sense.